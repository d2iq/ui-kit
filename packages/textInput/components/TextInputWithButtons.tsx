import * as React from "react";
import { cx } from "@emotion/css";
import TextInputWithIcon, { TextInputWithIconProps } from "./TextInputWithIcon";
import FormFieldWrapper from "../../shared/components/FormFieldWrapper";
import { flex, padding, flexItem } from "../../shared/styles/styleUtils";
import {
  inputContainer,
  getInputAppearanceStyle
} from "../../shared/styles/formStyles";
import { TextInputButtonProps } from "../../textInputButton/components/TextInputButton";

export interface TextInputWithButtonsProps
  extends Omit<TextInputWithIconProps, "iconEnd"> {
  /**
   * An array of TextInputButton components to render at the end of the input
   */
  buttons: Array<React.ReactElement<TextInputButtonProps>>;
}

class TextInputWithButtons extends TextInputWithIcon<
  TextInputWithButtonsProps
> {
  protected getInputElementProps() {
    const baseProps = super.getInputElementProps();
    const { buttons, ...inputProps } = baseProps as TextInputWithButtonsProps;

    return inputProps;
  }

  protected getInputContent() {
    const inputAppearance = this.getInputAppearance();

    return (
      <FormFieldWrapper
        // TODO: figure out how to get rid of this non-null assertion
        // If we stop generating an `id` prop in the TextInput component,
        // it would be possible for `this.props.id` to be undefined
        id={this.props.id!}
        errors={this.props.errors}
        hintContent={this.props.hintContent}
      >
        {({ getValidationErrors, getHintContent, isValid, describedByIds }) => (
          <React.Fragment>
            <div
              className={cx(
                flex(),
                padding("horiz", "s"),
                inputContainer,
                getInputAppearanceStyle(inputAppearance)
              )}
            >
              {this.getIconStartContent()}
              {this.getInputElement(
                [flexItem("grow"), padding("all", "none")],
                isValid,
                describedByIds
              )}
              {this.getButtons()}
            </div>
            {getHintContent}
            {getValidationErrors}
          </React.Fragment>
        )}
      </FormFieldWrapper>
    );
  }

  private getButtons() {
    if (!this.props.buttons.filter(Boolean)) {
      return;
    }

    return this.props.buttons.map((button, i) => (
      // TODO: consider making a component for this wrapper span
      <span
        className={cx(
          flex({ align: "center", justify: "center" }),
          flexItem("shrink"),
          { [padding("left", "s")]: i !== 0 }
        )}
        key={(React.isValidElement(button) && button.key) || i}
      >
        {button}
      </span>
    ));
  }
}

export default TextInputWithButtons;
