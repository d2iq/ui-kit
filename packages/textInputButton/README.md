### TextInputButton

This component is only used to pass buttons to the `TextInputWithButtons` component. Please see the documentation for the text input components for more info.
