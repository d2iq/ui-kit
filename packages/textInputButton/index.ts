export { default as TextInputButton } from "./components/TextInputButton";
