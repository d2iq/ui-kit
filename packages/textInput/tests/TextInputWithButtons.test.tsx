import React from "react";
import serializer from "@emotion/jest";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";

import TextInputWithButtons from "../components/TextInputWithButtons";
import { SystemIcons } from "../../icons/dist/system-icons-enum";
import { TextInputButton } from "../../textInputButton";

expect.addSnapshotSerializer(serializer);

describe("TextInputWithIcon", () => {
  it("renders", () => {
    const component = mount(
      <TextInputWithButtons
        id="renders"
        inputLabel="Label"
        buttons={[
          <TextInputButton key={0} shape={SystemIcons.Close} />,
          <TextInputButton key={1} shape={SystemIcons.Funnel} />
        ]}
      />
    );
    expect(toJson(component)).toMatchSnapshot();
  });
  it("renders with colored button", () => {
    const component = mount(
      <TextInputWithButtons
        id="renders.coloredBtn"
        inputLabel="Label"
        buttons={[
          <TextInputButton key={0} shape={SystemIcons.Close} color="red" />
        ]}
      />
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("calls function passed to button", () => {
    const buttonFn = jest.fn();
    const component = mount(
      <TextInputWithButtons
        id="callsFn"
        inputLabel="Label"
        buttons={[
          <TextInputButton
            key={0}
            shape={SystemIcons.Close}
            onClick={buttonFn}
          />
        ]}
      />
    );

    expect(buttonFn).not.toHaveBeenCalled();
    component.find("button").simulate("click");
    expect(buttonFn).toHaveBeenCalled();
  });
});
