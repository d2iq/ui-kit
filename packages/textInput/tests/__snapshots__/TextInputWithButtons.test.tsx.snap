// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputWithIcon renders 1`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

.emotion-1 {
  -webkit-align-items: flex;
  -webkit-box-align: flex;
  -ms-flex-align: flex;
  align-items: flex;
  height: auto;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  padding-left: 12px;
  padding-right: 12px;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  height: 36px;
  font-size: inherit;
  padding-left: 16px;
  padding-right: 16px;
  background-color: var(--themeBgPrimary, #FFFFFF);
  border-color: var(--themeBorder, #DADDE2);
}

.emotion-1>div {
  width: auto;
}

.emotion-1::-webkit-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1::-moz-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1:-ms-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1::placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input {
  font-size: inherit;
  color: inherit;
}

.emotion-1 input::-webkit-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input::-moz-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input:-ms-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input::placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1:focus {
  background-color: rgba(125,88,255,0.05);
  border-color: var(--themeBrandPrimary, #7D58FF);
}

.emotion-2 {
  background: none;
  border: 0;
  box-shadow: none;
  padding: 0;
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
  padding: 0;
}

.emotion-2:focus {
  outline: 0;
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
}

.emotion-3>div {
  width: auto;
}

.emotion-4 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  cursor: pointer;
}

.emotion-4::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-4>div:focus {
  outline: none;
}

.emotion-4:focus>div {
  outline-color: Highlight;
  outline-width: thin;
}

@media (-webkit-min-device-pixel-ratio: 0) {
  .emotion-4:focus>div {
    outline-color: -webkit-focus-ring-color;
    outline-style: auto;
    outline-width: unset;
  }
}

.emotion-5 {
  vertical-align: middle;
  fill: currentColor;
}

.emotion-5 use {
  pointer-events: none;
}

.emotion-6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  padding-left: 12px;
}

.emotion-6>div {
  width: auto;
}

<TextInputWithButtons
  appearance="standard"
  buttons={
    Array [
      <TextInputButton
        shape="system-close"
      />,
      <TextInputButton
        shape="system-funnel"
      />,
    ]
  }
  id="renders"
  inputLabel="Label"
  showInputLabel={true}
  type="text"
>
  <div
    data-cy="textInput textInput.standard"
  >
    <label
      className="emotion-0"
      data-cy="textInput-label"
      htmlFor="renders"
    >
      Label
    </label>
    <FormFieldWrapper
      id="renders"
    >
      <div
        className="emotion-1"
      >
        <input
          aria-describedby=""
          aria-invalid={false}
          className="emotion-2"
          data-cy="textInput-input"
          id="renders"
          onBlur={[Function]}
          onFocus={[Function]}
          type="text"
        />
        <span
          className="emotion-3"
          key="0"
        >
          <TextInputButton
            key="0"
            shape="system-close"
          >
            <ResetButton>
              <button
                className="emotion-4"
                type="button"
              >
                <div
                  tabIndex={-1}
                >
                  <Icon
                    shape="system-close"
                    size="xs"
                  >
                    <svg
                      aria-label="system-close icon"
                      className="emotion-5"
                      data-cy="icon"
                      height={16}
                      preserveAspectRatio="xMinYMin meet"
                      role="img"
                      viewBox="0 0 16 16"
                      width={16}
                    >
                      <use
                        xlinkHref="#system-close"
                      />
                    </svg>
                  </Icon>
                </div>
              </button>
            </ResetButton>
          </TextInputButton>
        </span>
        <span
          className="emotion-6"
          key="1"
        >
          <TextInputButton
            key="1"
            shape="system-funnel"
          >
            <ResetButton>
              <button
                className="emotion-4"
                type="button"
              >
                <div
                  tabIndex={-1}
                >
                  <Icon
                    shape="system-funnel"
                    size="xs"
                  >
                    <svg
                      aria-label="system-funnel icon"
                      className="emotion-5"
                      data-cy="icon"
                      height={16}
                      preserveAspectRatio="xMinYMin meet"
                      role="img"
                      viewBox="0 0 16 16"
                      width={16}
                    >
                      <use
                        xlinkHref="#system-funnel"
                      />
                    </svg>
                  </Icon>
                </div>
              </button>
            </ResetButton>
          </TextInputButton>
        </span>
      </div>
    </FormFieldWrapper>
  </div>
</TextInputWithButtons>
`;

exports[`TextInputWithIcon renders with colored button 1`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

.emotion-1 {
  -webkit-align-items: flex;
  -webkit-box-align: flex;
  -ms-flex-align: flex;
  align-items: flex;
  height: auto;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  padding-left: 12px;
  padding-right: 12px;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  height: 36px;
  font-size: inherit;
  padding-left: 16px;
  padding-right: 16px;
  background-color: var(--themeBgPrimary, #FFFFFF);
  border-color: var(--themeBorder, #DADDE2);
}

.emotion-1>div {
  width: auto;
}

.emotion-1::-webkit-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1::-moz-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1:-ms-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1::placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input {
  font-size: inherit;
  color: inherit;
}

.emotion-1 input::-webkit-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input::-moz-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input:-ms-input-placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1 input::placeholder {
  color: var(--themeTextColorPrimary, #1B2029);
  opacity: 0.4;
}

.emotion-1:focus {
  background-color: rgba(125,88,255,0.05);
  border-color: var(--themeBrandPrimary, #7D58FF);
}

.emotion-2 {
  background: none;
  border: 0;
  box-shadow: none;
  padding: 0;
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
  padding: 0;
}

.emotion-2:focus {
  outline: 0;
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
}

.emotion-3>div {
  width: auto;
}

.emotion-4 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  cursor: pointer;
}

.emotion-4::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-4>div:focus {
  outline: none;
}

.emotion-4:focus>div {
  outline-color: Highlight;
  outline-width: thin;
}

@media (-webkit-min-device-pixel-ratio: 0) {
  .emotion-4:focus>div {
    outline-color: -webkit-focus-ring-color;
    outline-style: auto;
    outline-width: unset;
  }
}

.emotion-5 {
  vertical-align: middle;
  fill: red;
}

.emotion-5 use {
  pointer-events: none;
}

<TextInputWithButtons
  appearance="standard"
  buttons={
    Array [
      <TextInputButton
        color="red"
        shape="system-close"
      />,
    ]
  }
  id="renders.coloredBtn"
  inputLabel="Label"
  showInputLabel={true}
  type="text"
>
  <div
    data-cy="textInput textInput.standard"
  >
    <label
      className="emotion-0"
      data-cy="textInput-label"
      htmlFor="renders.coloredBtn"
    >
      Label
    </label>
    <FormFieldWrapper
      id="renders.coloredBtn"
    >
      <div
        className="emotion-1"
      >
        <input
          aria-describedby=""
          aria-invalid={false}
          className="emotion-2"
          data-cy="textInput-input"
          id="renders.coloredBtn"
          onBlur={[Function]}
          onFocus={[Function]}
          type="text"
        />
        <span
          className="emotion-3"
          key="0"
        >
          <TextInputButton
            color="red"
            key="0"
            shape="system-close"
          >
            <ResetButton>
              <button
                className="emotion-4"
                type="button"
              >
                <div
                  tabIndex={-1}
                >
                  <Icon
                    color="red"
                    shape="system-close"
                    size="xs"
                  >
                    <svg
                      aria-label="system-close icon"
                      className="emotion-5"
                      data-cy="icon"
                      height={16}
                      preserveAspectRatio="xMinYMin meet"
                      role="img"
                      viewBox="0 0 16 16"
                      width={16}
                    >
                      <use
                        xlinkHref="#system-close"
                      />
                    </svg>
                  </Icon>
                </div>
              </button>
            </ResetButton>
          </TextInputButton>
        </span>
      </div>
    </FormFieldWrapper>
  </div>
</TextInputWithButtons>
`;
